### PR TITLE
feat: add AccountHashingStage and StorageHashingStage

### PR DESCRIPTION
**Pull request by rakita**
Adds AccountHashingStage and StorageHashingStage, these stages are preparation for merkle tree root calculation and allows us to read storage/account hash them all and in ordered way insert them in `HashingAccount/HashingStorage` tables.

These stages has two modes of execution. First mode will drop that hashing tables and iterate over whole state to hash it, and second mode will read changeset and apply diff to hashing tables. Unwind does it only on changesets as unwind in most cases will be only on a few blocks.  

Removed hash from `BLockNumHash` for block transition table.

Added empty `MerkleStage`

### CODE REVIEW

**Comment by onbjerg** [comment]
Does this not clash with #704?

**Comment by rakita** [comment]
> Does this not clash with #704?

There were some clashes as I was on vecation, see here: https://github.com/paradigmxyz/reth/issues/634

This is panic free version,  and Hash from `NumHash` from `BlockTransitionIndex`  is removed.

**Comment by gakonst** [comment]
@rakita anything missing from this PR to get it in?

**Comment by rakita** [comment]
> @rakita anything missing from this PR to get it in?

It is ready to go in.

File: crates/storage/provider/src/db_provider/storage.rs [hunk 1]
```diff
@@ -46,11 +41,18 @@ impl<DB: Database> StateProviderFactory for ProviderImpl<DB> {
         let block_number =
             tx.get::<tables::HeaderNumbers>(block_hash)?.ok_or(Error::BlockHash { block_hash })?;
 
+        // check if block is canonical or not. Only canonical blocks have changesets.
+        let canonical_block_hash = tx
```
Comment (gakonst) [inline crates/storage/provider/src/db_provider/storage.rs:45]: Was this a bug / missing check? Let's pull in separate PR and a test for this case if so?

File: crates/storage/provider/src/block.rs [hunk 2]
```diff
@@ -168,15 +166,25 @@ pub fn insert_canonical_block<'a, TX: DbTxMut<'a> + DbTx<'a>>(
         let rec_tx = transaction.clone().into_ecrecovered().unwrap();
         tx.put::<tables::TxSenders>(current_tx_id, rec_tx.signer())?;
         tx.put::<tables::Transactions>(current_tx_id, rec_tx.into())?;
+        transition_id += 1;
```
Comment (gakonst) [inline crates/storage/provider/src/block.rs:169]: Was this a bug? We had to increment it before writing?

File: crates/storage/provider/src/block.rs [hunk 3]
```diff
@@ -134,28 +135,25 @@ pub fn insert_canonical_block<'a, TX: DbTxMut<'a> + DbTx<'a>>(
...
-            let last_transition_id = tx
-                .get::<tables::BlockTransitionIndex>((prev_block_num, prev_block_hash).into())?
-                .ok_or(ProviderError::BlockTransition {
-                    block_number: prev_block_num,
-                    block_hash: prev_block_hash,
-                })?;
-            (prev_body.start_tx_id + prev_body.tx_count, last_transition_id + 1)
-        }
+    let (mut current_tx_id, mut transition_id) = if let Some(embed) = parent_tx_num_transition_id {
+        embed
```
Comment (gakonst) [inline crates/storage/provider/src/block.rs:139]: What does embed mean here?

File: crates/storage/provider/src/block.rs [hunk 4]
```diff
@@ -117,10 +117,11 @@ pub struct ChainInfo {
 /// Inserts blocks data to [tables::CanonicalHeaders], [tables::Headers], [tables::HeaderNumbers],
 /// and transactions data to [tables::TxSenders], [tables::Transactions],
 /// [tables::CumulativeTxCount] and [tables::BlockBodies]
-pub fn insert_canonical_block<'a, TX: DbTxMut<'a> + DbTx<'a>>(
```
Comment (gakonst) [inline crates/storage/provider/src/block.rs:120]: Was it a mistake to call it canonical before?

File: crates/storage/db/src/tables/mod.rs [hunk 5]
```diff
@@ -254,6 +260,22 @@ dupsort!(
...
+    /// We are saving whole account data as it is needed for partial update when
+    /// part of storage is changed. Benefit for merkelization is that hashed addresses are sorted.
+    ( HashedAccount ) H256 | Account
+);
+
+dupsort!(
+    /// Stores the current storage values indexed with `keccak256(Address)` and
+    /// hash of storage key `keccak256(key)`.
+    /// This table is in preparation for merkelization and calculation of state root.
+    /// Benefit for merkelization is that hashed addresses/keys are sorted.
```
Comment (gakonst) [inline crates/storage/db/src/tables/mod.rs:275]: How are they sorted after hashing? A bit confusing comment.

Also 
```suggestion
    /// Benefit for merklization is that hashed addresses/keys are sorted.
```

File: crates/storage/db/src/tables/mod.rs [hunk 6]
```diff
@@ -176,14 +173,23 @@ table!(
 
 table!(
     /// Stores the mapping of block number to state transition id.
-    ( BlockTransitionIndex ) BlockNumHash | TransitionId
+    /// The block transition marks the final state at the end of the block.
+    /// Increment the transition if the block contains an addition block reward.
+    /// If the block does not have a reward and transaction, the transition will be the same as the
+    /// transition at the last transaction of this block.
+    ( BlockTransitionIndex ) BlockNumber | TransitionId
```
Comment (gakonst) [inline crates/storage/db/src/tables/mod.rs:180]: Would prefer we do this change in a standalone PR, probably needs to be part of our broader change to only using block numbers? @onbjerg @joshieDo

File: crates/stages/src/test_utils/macros.rs [hunk 7]
```diff
@@ -97,7 +97,7 @@ macro_rules! stage_test_suite {
 
             // Run stage unwind
             let unwind_input = crate::stage::UnwindInput {
-                unwind_to: stage_progress, stage_progress, bad_block: None,
+                unwind_to: stage_progress, stage_progress: previous_stage, bad_block: None,
```
Comment (gakonst) [inline crates/stages/src/test_utils/macros.rs:100]: Was this a bug? Cc @rkrasiuk

File: crates/stages/src/stages/merkle.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+use std::fmt::Debug;
+use tracing::*;
+
+const MERKLE_EXECUTION: StageId = StageId("MerkleExecuteStage");
+const MERKLE_UNWIND: StageId = StageId("MerkleUnwindStage");
+
+/// Merkle stage uses input from [AccountHashingStage] and [StorageHashingStage] stages
+/// and calculated intermediate hashed and state root.
+/// As input of this stage depends on Hashing stages this stage will be for execute be put after
+/// but for unwind before the hashing stages.
```
Comment (gakonst) [inline crates/stages/src/stages/merkle.rs:14]:
```suggestion
/// This stage depends on the Account and Storage stages. It will be executed after them during execution, and before them during unwinding.
```

File: crates/stages/src/stages/hashing_storage.rs [hunk 9]
```diff
@@ -0,0 +1,491 @@
...
+            }
+            storage_cursor.append_dup(tid_address.address(), entry)?;
+
+            tx.cursor_dup_write::<tables::StorageChangeSet>()?
+                .append_dup(tid_address, prev_entry)?;
+            Ok(())
+        }
+
+        fn unwind_storage(&self, input: UnwindInput) -> Result<(), TestRunnerError> {
+            println!("\n unwinding storage...");
```
Comment (gakonst) [inline crates/stages/src/stages/hashing_storage.rs:461]:
```suggestion
            tracing::debug!("unwinding storage...");
```

**Review by gakonst — CHANGES_REQUESTED** [review]
Haven't gone through the logic inside the 2 hashing stages, looked at all other files and would like separate PRs for the driveby changes like the TransitionId key and the canonical block so we keep things clean. Will take a closer look at the actual hashing stage implementations soon

File: crates/cli/utils/src/init.rs [hunk 10]
```diff
@@ -52,7 +52,7 @@ pub fn init_genesis<DB: Database>(db: Arc<DB>, genesis: Genesis) -> Result<H256,
     tx.put::<tables::CanonicalHeaders>(0, hash)?;
     tx.put::<tables::HeaderNumbers>(hash, 0)?;
     tx.put::<tables::BlockBodies>((0, hash).into(), Default::default())?;
-    tx.put::<tables::BlockTransitionIndex>((0, hash).into(), 0)?;
+    tx.put::<tables::BlockTransitionIndex>(0, 0)?;
```
[the remark this answers is not in the document]
Comment (rakita) [reply crates/cli/utils/src/init.rs:55]: I agree

File: crates/stages/src/test_utils/macros.rs (hunk 7 above)
Comment (rakita) [reply crates/stages/src/test_utils/macros.rs:100]: Yes, it was, but in the way that tests were set it didn't show.

File: crates/storage/provider/src/block.rs (hunk 4 above)
Comment (rakita) [reply crates/storage/provider/src/block.rs:120]: insert_canonical_block function is still present below. This is mostly used for tests

File: crates/storage/provider/src/block.rs (hunk 3 above)
Comment (rakita) [reply crates/storage/provider/src/block.rs:139]: hardcoded value for parent tx num and transition id.

File: crates/storage/provider/src/block.rs (hunk 2 above)
Comment (rakita) [reply crates/storage/provider/src/block.rs:169]: Order was changed, before it was incremented at lines above.

File: crates/storage/provider/src/db_provider/storage.rs (hunk 1 above)
Comment (rakita) [reply crates/storage/provider/src/db_provider/storage.rs:45]: This is result of removing hash from  `BlockTransition` (Index was `BlockNumHash`) table. It is better to be sure if given hash is in canonical block or not

File: crates/storage/db/src/tables/mod.rs (hunk 6 above)
Comment (rakita) [reply crates/storage/db/src/tables/mod.rs:180]: Asked in chat, but will move it to separate PR for a cleaner look. It is already changed everywhere it is used.

File: crates/stages/src/test_utils/macros.rs (hunk 7 above)
Comment (rkrasiuk) [reply crates/stages/src/test_utils/macros.rs:100]: @rakita thanks for spotting & fixing

Comment (rakita) [reply crates/stages/src/test_utils/macros.rs:100]: moved change here: https://github.com/paradigmxyz/reth/pull/895

File: crates/storage/db/src/tables/mod.rs (hunk 6 above)
Comment (rakita) [reply crates/storage/db/src/tables/mod.rs:180]: Split the change to separate PR here: https://github.com/paradigmxyz/reth/pull/894

File: crates/stages/src/stages/merkle.rs [hunk 11]
```diff
@@ -0,0 +1,64 @@
+use crate::{
```
Comment (gakonst) [inline crates/stages/src/stages/merkle.rs:1]: Probably can leave this stage for separate PR given its a noop right now?

File: crates/stages/src/stages/hashing_storage.rs [hunk 12]
```diff
@@ -0,0 +1,491 @@
...
+                let start_block = input.stage_progress.unwrap_or_default() + 1;
+                let end_block = output.stage_progress;
+                if start_block > end_block {
+                    return Ok(())
+                }
+            }
+            self.check_hashed_storage()
+        }
+
+        async fn after_execution(&self, _seed: Self::Seed) -> Result<(), TestRunnerError> {
```
Comment (gakonst) [inline crates/stages/src/stages/hashing_storage.rs:372]: Is this used only for debugging or do we also want to add any asserts here?

File: crates/stages/src/stages/hashing_storage.rs [hunk 13]
```diff
@@ -0,0 +1,491 @@
...
+                            "{expected}: {address:?}"
+                        );
+                        expected += 1;
+                    }
+                    let count = tx
+                        .cursor_dup_read::<tables::HashedStorage>()?
+                        .walk([0; 32].into())?
+                        .count();
+
+                    assert_eq!(count, expected);
```
Comment (gakonst) [inline crates/stages/src/stages/hashing_storage.rs:426]: Sweet.

File: crates/stages/src/stages/hashing_storage.rs [hunk 14]
```diff
@@ -0,0 +1,491 @@
...
+        &mut self,
+        tx: &mut Transaction<'_, DB>,
+        input: UnwindInput,
+    ) -> Result<UnwindOutput, StageError> {
+        let from_transition_rev = tx.get_block_transition(input.unwind_to)? + 1;
+        let to_transition_rev = tx.get_block_transition(input.stage_progress)? + 1;
+
+        let mut hashed_storage = tx.cursor_dup_write::<tables::HashedStorage>()?;
+
+        // Aggregate all transition changesets and and make list of account that have been changed.
```
Comment (gakonst) [inline crates/stages/src/stages/hashing_storage.rs:167]:
```suggestion
        // Aggregate all transition changesets and make list of accounts that have been changed.
```

File: crates/stages/src/stages/hashing_storage.rs [hunk 15]
```diff
@@ -0,0 +1,491 @@
...
+        tx.cursor_read::<tables::StorageChangeSet>()?
+            .walk((from_transition_rev, H160::zero()).into())?
+            .take_while(|res| {
+                res.as_ref()
+                    .map(|(TransitionIdAddress((k, _)), _)| *k < to_transition_rev)
+                    .unwrap_or_default()
+            })
```
Comment (gakonst) [inline crates/stages/src/stages/hashing_storage.rs:174]: Would like to figure out a good helper for walking+take while, maybe like a range(..) helper generic over the key?

File: crates/stages/src/stages/hashing_storage.rs [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+
+                    // next key of iterator
+                    let next_key = storage.next()?;
+
+                    let mut hashes = tx.cursor_write::<tables::HashedStorage>()?;
+                    // iterate and append presorted hashed slots
+                    hashed_batch.into_iter().try_for_each(|(k, v)| hashes.append(k, v))?;
+                    next_key
+                };
+                tx.commit()?;
```
Comment (gakonst) [inline crates/stages/src/stages/hashing_storage.rs:82]: Do we want to be committing inside the stage? Thought we want to avoid commits internally and return control flow to the pipeline once we hit commit threshold?

That would also mean we could remove the loop above?

File: crates/stages/src/stages/hashing_storage.rs [hunk 17]
```diff
@@ -0,0 +1,491 @@
...
+        // read storage changeset, merge it into one changeset and calculate storage hashes.
+        let from_transition = tx.get_block_transition(stage_progress)? + 1;
+        let to_transition = tx.get_block_transition(previous_stage_progress)? + 1;
+
+        // if there are more blocks then threshold it is faster to go over Plain state and hash all
+        // account otherwise take changesets aggregate the sets and apply hashing to
+        // AccountHashing table
+        if to_transition - from_transition > self.clean_threshold {
+            tx.clear::<tables::HashedStorage>()?;
+            tx.commit()?;
```
Comment (gakonst) [inline crates/stages/src/stages/hashing_storage.rs:54]: How does Erigon do this? Do they drop the table every time?

File: crates/stages/src/stages/hashing_storage.rs [hunk 18]
```diff
@@ -0,0 +1,491 @@
...
+        let stage_progress = input.stage_progress.unwrap_or_default();
+        let previous_stage_progress = input.previous_stage_progress();
+
+        // read storage changeset, merge it into one changeset and calculate storage hashes.
+        let from_transition = tx.get_block_transition(stage_progress)? + 1;
+        let to_transition = tx.get_block_transition(previous_stage_progress)? + 1;
+
+        // if there are more blocks then threshold it is faster to go over Plain state and hash all
+        // account otherwise take changesets aggregate the sets and apply hashing to
+        // AccountHashing table
```
Comment (gakonst) [inline crates/stages/src/stages/hashing_storage.rs:51]: Why is this true?

**Review by gakonst — CHANGES_REQUESTED** [review]
Great work.

Hashing logic in both stages LGTM. Test scaffolds as well LGTM. 

Would like to remove the loops and the TX.commit inside the stages and instead rely on the pipeline to commit for us.

Main Q on my mind is how we'll be able to call these stages repeatedly to build the state root without repeating a ton of work.

File: crates/stages/src/stages/merkle.rs (hunk 11 above)
Comment (rakita) [reply crates/stages/src/stages/merkle.rs:1]: It is there for strange logic of execution/unwind, I am afraid it would get lost if removed and it is easier to have it right here.

File: crates/stages/src/stages/hashing_storage.rs (hunk 15 above)
Comment (rakita) [reply crates/stages/src/stages/hashing_storage.rs:174]: it will be benefitial! This pattern is used in most of the stages.

File: crates/stages/src/stages/hashing_storage.rs (hunk 16 above)
Comment (rakita) [reply crates/stages/src/stages/hashing_storage.rs:82]: Yes, if we return, there is no way to save progress. Here, we are iterating over all keys in the table, not over the blocks as in other stages.

File: crates/stages/src/stages/hashing_storage.rs (hunk 17 above)
Comment (rakita) [reply crates/stages/src/stages/hashing_storage.rs:54]: Not every time, but if a lot of blocks are executed at once. It made sense for me as applying every changetset would take a lot of time. Erigon does in intermediatehash stage: https://github.com/ledgerwatch/erigon/blob/937fb2c4d95b0640904b7dd715c1736e7d0a5ede/eth/stagedsync/stage_interhashes.go#L149-L150

File: crates/stages/src/stages/hashing_storage.rs (hunk 18 above)
Comment (rakita) [reply crates/stages/src/stages/hashing_storage.rs:51]: Changesets over time can be more then just end result. Just imagine one ERC20 balance changed every block

File: crates/stages/src/stages/hashing_storage.rs (hunk 12 above)
Comment (rakita) [reply crates/stages/src/stages/hashing_storage.rs]: Yeah, there were debugs only prints, removed them.

**Comment by rakita** [comment]
@gakonst this is ready to go in

**Review by gakonst — APPROVED** [review]
As discussed, feeling a bit uneasy on how performant this is going to be if it's called continuously, but we will understand better once we get there. The hashing logic is needed and looks correct, so going to merge.
